msg_modification/tests_encode_to_chunked.py

```diff
@@ -225,3 +252,498 @@ def test_transform_possible(self):
+        self.assertEqual(rec_resp.status, '200', 'Unexpected response')
+        self.assertEqual(rec_resp.headers.get('Transfer-Encoding', None),
+                         'chunked', 'Chunked encoding is missed!')
+        self.assertEqual(rec_resp.body, expected_body,
+                         'Body is not in chunked encoding or mismatch!')
+
+    def test_zero_body_to_chunked(self):
+        """Server doesn't send a body (zero sized body). Tempesta converts
+        the message to chunked encoding.
+        """
+        self.start_all()
+        client = self.get_client('deproxy')
+        srv = self.get_server('deproxy')
+
+        date = deproxy.HttpMessage.date_time_string()
+        body = self.gen_body(0)
+        chunked_body = self.chunked_body(body)
+        expected_body = self.chunked_body(body, lowercase=True)
+        resp = ('HTTP/1.1 200 OK\r\n'
+                'Connection: keep-alive\r\n'
+                'Content-type: text/html\r\n'
+                'Last-Modified: Mon, 12 Dec 2016 13:59:39 GMT\r\n'
+                'Server: Deproxy Server\r\n'
+                'Transfer-Encoding: chunked\r\n'
+                'Date: %s\r\n'
+                '\r\n%s'
+                % (date, chunked_body))
+        srv.set_response(resp)
+
+        rec_resp = self.send_req(client)
+        self.assertFalse(client.connection_is_closed(),
+                         'Client connection was unexpectedly closed')
+        self.assertEqual(rec_resp.status, '200', 'Unexpected response')
+        self.assertEqual(rec_resp.headers.get('Transfer-Encoding', None),
+                         'chunked', 'Chunked encoding is missed!')
+        self.assertEqual(rec_resp.body, expected_body,
+                         'Body is not in chunked encoding or mismatch!')
+
+    def test_trailer_headers(self):
+        """ Server sends chunked trailer.
+        Tempesta strips the chunked encoding and applies it back, but message
+        looks absolutely identical and trailer header sits on it's place.
+        """
+        self.start_all()
+        client = self.get_client('deproxy')
+        srv = self.get_server('deproxy')
+
+        date = deproxy.HttpMessage.date_time_string()
+        body = self.gen_body(40)
+        trailer = 'X-My-trailer-Header: somevalue'
+        chunked_body = self.chunked_body_with_trailer(body, trailer)
+        # self.chunked_body() adds extra CRLF after '0' chunk descriptor,
+        # not needed here because of chunked trailer
+        expected_body = '\r\n'.join(["%x" % len(body), body, '0', ''])
+        resp = ('HTTP/1.1 200 OK\r\n'
+                'Connection: keep-alive\r\n'
+                'Content-type: text/html\r\n'
+                'Last-Modified: Mon, 12 Dec 2016 13:59:39 GMT\r\n'
+                'Server: Deproxy Server\r\n'
+                'Transfer-Encoding: chunked\r\n'
+                'Date: %s\r\n'
+                '\r\n%s\r\n'
+                % (date, chunked_body))
+        srv.set_response(resp)
```

Comment:
It seems the trailer headers tests are implemented for HTTP/1 clients only, but #1410.2 mentions the problem about caching responses with trailer headers. Please implement trailer headers tests for h2 as well. At least need to test (1) responses with trailer headers directly proxied to an h2 client and (2) response with trailed header is firstly cached by h1 request and next returned to an h2 client.